### PR TITLE
WebSockets support

### PR DESCRIPTION
**Pull request by damianh**
Adds support for websocket proxying.

Usage looks like

```csharp
app.UseWebSocketProxy(new Uri($"ws://upstream:port/"));
```

Unfortunately, because TestServer contains it's own `WebSocketClient` and doesn't use the real `ClientWebSocket`, there isn't a story for embedded testing via the router.

### CODE REVIEW

**Comment by Kormyr** [comment]
Hi,

First thanks for this library and the effort you put into it.
I tried this websocket proxy feature and have a few questions:

1) I could not get it to work, how are we supposed to use it apart from the line `app.UseWebSocketProxy(new Uri("wss://myserver.com"));` line? When I check browser requests I see that there is an issue on the handshake.

2) When is it planned to release version 2.1.0?

PS: I did not create an issue yet because it is not a released feature yet, let me know if I should create one instead.

**Comment by samcov** [comment]
I'm having the same issue where it initially connects then disconnects with handshake errors. I'm using SignalR for websocket support, attached is the normal version and the failed reverse proxy version.
ReverseProxy1
ReverseProxy2

**Comment by damianh** [comment]
Ok seems to be an issue here. Will investigate before releasing.

**Comment by samcov** [comment]
Thanks!

**Comment by damianh** [comment]
@samcov I believe I have at least reproduced the error. Opened a failing PR: https://github.com/damianh/ProxyKit/pull/89 . A quick 👀 would be appreciated to see if it is same thing you are experiencing.

@Kormyr SignalR uses `http(s)://` and `ws(s)://` thus under ProxyKit current design you need to proxy both protocols as you can see here: https://github.com/damianh/ProxyKit/pull/89/files#diff-70826f71dadcf32a3661f348f2ff3ddbR125

I am going to reflecting on the design of this now. 2.1.0 will ship when I am happier with it (and works fully!)

**Comment by damianh** [comment]
@samcov @Kormyr I've added two recipes - one for websockets and one for singalr (and I have a signalr specific test).

The handshake canceled issues is because the proxy itself needs to be configured to handle incoming web socket upgrade requests first before forwarding it. See https://github.com/damianh/ProxyKit/blob/master/src/Recipes/15_WebSockets.cs

Also, for signalr specific concerns, you need to forward signalr `http` and `ws` traffic. See https://github.com/damianh/ProxyKit/blob/master/src/Recipes/16_SignalR.cs

Note: to get this working I didn't have to change ProxyKit code.

However, while this works (well appears to from tests 😄), I'm not 100% happy with the design of it. I'm going to have a chew on it shortly before releasing.

Finally, thank you both for taking the time to try this out and give feedback. It has been very valuable.

**Comment by markusdresch** [comment]
maybe you want to add that the full hub url has to be proxied for signalr, not only the upstream server. so for example it's `app.UseWebSocketProxy(new Uri("ws://localhost:5000/signalr"))`

**Comment by markusdresch** [comment]
i am also passing parameters to the signalr url. changed WebSocketProxyMiddleware on line 75:
```
var full = new Uri(destinationUri.OriginalString + context.Request.Path + context.Request.QueryString);
await client.ConnectAsync(full, context.RequestAborted).ConfigureAwait(false);
```

**Comment by samcov** [comment]
image
The latest code fails to even connect. I also had to add a CORS policy because of a cross site scripting issue, hopefully that isn't the reason I can't connect. Although I didn't need a CORS policy in the previous version, which still works but disconnects.

**Comment by bioharz** [comment]
Maybe this pull request can help:
https://github.com/damianh/ProxyKit/pull/98
I build my own middleware and proxy server and I have some similar issues with SignalR...

**Comment by damianh** [comment]
Hi, I've pushed 2.1.0-beta.1 to nuget.org that should fix this. I would sincerely  appreciate if you could kick the tires and see if it works for you. TIf it does, I'll release it.

Thank you for your patience:)

**Comment by samcov** [comment]
image
Firstly, **THIS WORKS**!!!

Here are a couple of housekeeping things I noticed.
1. The project was missing the code in program.cs, I had to add in the menuing system to select
2. The abstract Recipe class with the Run method was also missing

The code works, however, there is a delay on startup as you can see in the console,
1. There is a failure in the initial handshake, and then
2. It fails to start the transport with Websockets being undefined

However, after that, it rights itself and works a treat!

What you see in the picture is  our application which uses SignalR to get everything except the html. It uses SignalR to grab scripts from RavenDB so that users may edit or run them using Roslyn, basically building functions in the cloud, so it's working very well.

I think there is a timeout issue that causes the initial delay connection due to the mentioned errors.

**Comment by damianh** [comment]
> Firstly, THIS WORKS!!!

Great! 😄 

> The abstract Recipe class with the Run method was also missing 

I removed it because I want the recipes to be clean and focused. The abstract base class was a distraction. I indent to add runable examples to the repo as separate projects.

**Comment by damianh** [comment]
Actually 🤔 ... the screen shot shows that SignalR is falling back to SSE (Server sent events) and not actually using websocket though. That's not desirable. Hmm, need to check that....

**Comment by samcov** [comment]
> Actually 🤔 ... the screen shot shows that SignalR is falling back to SSE (Server sent events) and not actually using websocket though. That's not desirable. Hmm, need to check that....

**You're right**, not quite what we want.

**Comment by damianh** [comment]
@samcov I've added a runnable SignalR simple chat application to the solution and it seems to be working fine.

image

I've even made some tweaks for another issue [regarding pathing](https://github.com/damianh/ProxyKit/commit/227e7849ecb4c997aa227d5f2c47701a82179290) that, from what I can see works fine.

For me to be able to diagnose any issues I really need to reproduce it. If you still think there is an issue, please tweak that sample app so it represents your configuration etc and shows the problem so that I can run it myself. 

Cheers.

**Comment by damianh** [comment]
If the issue can be reproduced, please create a new issue. Cheers ;)

**Comment by samcov** [comment]
Error when running SignalRSimpleChat:

System.IO.IOException
  HResult=0x80131620
  Message=Failed to bind to address http://127.0.0.1:40066: address already in use.
  Source=Microsoft.AspNetCore.Server.Kestrel.Core
  StackTrace:
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.<BindEndpointAsync>d__3.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions.<BindAsync>d__43.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.AddressesStrategy.<BindAsync>d__2.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.<BindAsync>d__0.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer.<StartAsync>d__21`1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.<StartAsync>d__26.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Hosting.WebHostExtensions.<RunAsync>d__5.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Hosting.WebHostExtensions.<RunAsync>d__4.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Hosting.WebHostExtensions.Run(IWebHost host)
   at ProxyKit.Recipe.SignalRSimpleChat.Program.<Main>d__0.MoveNext() in D:\Desktop\Testbed\ProxyKit-master\src\SignalRSimpleChat\Program.cs:line 23
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at ProxyKit.Recipe.SignalRSimpleChat.Program.<Main>(String[] args)

Inner Exception 1:
AddressInUseException: Only one usage of each socket address (protocol/network address/port) is normally permitted

Inner Exception 2:
SocketException: Only one usage of each socket address (protocol/network address/port) is normally permitted

**Comment by damianh** [comment]
>  http://127.0.0.1:40066: address already in use.

A) You have something else running on that port. There explanation is right there. 
B) How are you getting port 40066?? The port numbers in user are 5000 and 5001.

https://github.com/damianh/ProxyKit/blob/db4fc9274563e746838d8750c57c7f0f09d1bf13/src/SignalRSimpleChat/Program.cs#L11-L19

image

**Comment by samcov** [comment]
I'll check, but I tried it on two different machines with the same result.

**Comment by markusdresch** [comment]
you probably run it on iis express in debug mode. simply remove that debug configuration.

**Comment by samcov** [comment]
@markusdresch Yes, he has launch settings that conflict. I got busy and haven't had time to get back to it, but I will soon.

BTW, thanks for the comment!

**Comment by samcov** [comment]
image
Ok, finally got a break from customers whining, LOL! BTW, I'm using command line for less conflicts.

1. I modified your sample to eliminate your **Chat App**, and ONLY run the proxy listening on port 5000 as usual.

2. I have my SignalR application listening on port 7080, so I proxy ws://localhost:7080 & http://localhost:7080

3. I start the Web Pages on http://localhost:8081, it's pure HTML, but uses SignalR to talk to the service on the proxied port of 5000.

**Bottom Line**: This versions works perfectly!!!

In the picture you can see that it connects to the hub via websockets on port 5000, using hub protocol json, exactly the way it normally behaves, just through the proxy.

**Excellent Work!**

**Comment by damianh** [comment]
I've cleaned up the launchsettings.json files via #116 . It was gitignored and my local copy didn't have the IIS profile...

> Bottom Line: This versions works perfectly!!!

Fantastic news :)

Pending #115, I'll do a release soon.

Thanks for the feedback and assistance.

**Comment by damianh** [comment]
@samcov I've put out a `beta.2` package on nuget.org https://www.nuget.org/packages/ProxyKit/2.1.0-beta.2 . It has a small change because of #115 . I'd just like to validate with you that things still work before releasing. Cheers !

**Comment by samcov** [comment]
@damianh **Great Job**: Everything is working just fine with the new release on my end!

**Let her rip :)**
